File: 20231013_Friday/20231013.go
// Leetcode: 746. Min Cost Climbing Stairs.
// https://leetcode.com/problems/min-cost-climbing-stairs
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 5 ms, faster than 59.14% of Go online submissions for Min Cost
// Climbing Stairs.
// Memory Usage: 2.9 MB, less than 79.74% of Go online submissions for Min Cost
// Climbing Stairs.
// 2023.10.13 Daily Challenge.
package main

func minCostClimbingStairs(cost []int) int {
	n := len(cost)
	if 2 == n {
		return min(cost[0], cost[1])
	}
	for i := 2; i < n; i += 1 {
		cost[i] += min(cost[i-1], cost[i-2])
	}
	return min(cost[n-1], cost[n-2])
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: 20231013_Friday/20231013.rb
# Leetcode: 746. Min Cost Climbing Stairs.
# https://leetcode.com/problems/min-cost-climbing-stairs
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 76 ms, faster than 17.89% of Ruby online submissions for Min Cost
# Climbing Stairs.
# Memory Usage: 211.4 MB, less than 85.26% of Ruby online submissions for Min
# Cost Climbing Stairs.
# 2023.10.13 Daily Challenge.
# @param {Integer[]} cost
# @return {Integer}
def min_cost_climbing_stairs(cost)
  # Pull DP. Inplace solution.
  # Select minimum from i-1 and i-2, plus current.
  
  # If less then three elements no need to count.
  return cost.min if 2 >= cost.size

  
  (2...cost.size).each do |i|
    cost[i] = [cost[i-1],cost[i-2]].min + cost[i]
  end

  # Return minumum from two lest steps to reach the end.
  return [cost[-1], cost[-2]].min
end
